walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,83 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> requiredTimeRanges = findTimeSlotsForNewMeeting(
+        filterRelevantEvents(events, request.getAttendees()),
+        request.getDuration());
+    List<TimeRange> optionalTimeRanges = findTimeSlotsForNewMeeting(
+        filterRelevantEvents(events, request.getOptionalAttendees()),
+        request.getDuration());
+    List<TimeRange> intersection = findTimeRangesIntersections(
+        requiredTimeRanges, optionalTimeRanges, request.getDuration());
+    if (intersection.isEmpty()) {
+      return requiredTimeRanges;
+    }
+    return intersection;
+  }
+
+  private static List<TimeRange> findTimeRangesIntersections(List<TimeRange> required,
+                                                             List<TimeRange> optional,
+                                                             long duration) {
+    List<TimeRange> intersection = new ArrayList<>();
+    for (TimeRange requiredTimeRange : required) {
+      for (TimeRange optionalTimeRange : optional) {
+        createIntersection(requiredTimeRange, optionalTimeRange)
+            .filter(timeRange -> timeRange.duration() >= duration)
+            .ifPresent(intersection::add);
+      }
+    }
+    return intersection;
+  }
+
+  private static Optional<TimeRange> createIntersection(TimeRange a, TimeRange b) {
+    if (a.overlaps(b)) {
+      return Optional.of(TimeRange.fromStartEnd(
+          Math.max(a.start(), b.start()),
+          Math.min(a.end(), b.end()),
+          false));
+    }
+    return Optional.empty();
+  }
+
+  private static List<TimeRange> findTimeSlotsForNewMeeting(Collection<Event> events,
+                                                            long duration) {
+    List<Event> sortedEvents = sortEvents(events);
+    List<TimeRange> queryResult = new ArrayList<>();
+    int currentTime = 0;
+    for (Event event : sortedEvents) {
+      if (currentTime + duration <= event.getWhen().start()) {
+        int end = (int) Math.max(event.getWhen().start(), currentTime + duration);
+        queryResult.add(TimeRange.fromStartEnd(currentTime, end, false));
```

Comment:
you could use [fromStartDuration](https://github.com/LightFelicis/step/blob/81f3de8b3d122e9b04ec7846422031f39ce70947/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/TimeRange.java#L183).

Reply:
Done.